### ISSUE TITLE
How to get page number?

### ISSUE THREAD

**Issue by enrac5** [python-openxml/python-docx#1332]
Is there a way via the docx API to get the page number when I'm parsing a table?

**Comment by Mathacc** [comment]
Well, yes and no, depends on the engine the docx was created. The pages are actually renderized on the engine you are running your docx file, what I mean with that: for example, the pages can be different if you open a docx file on google docs or office package. But there is a way to get the possible pages that were created. You must search for something like this in your xml file:
```python
if 'lastRenderedPageBreak' in element._element.xml or ('w:br' in element._element.xml and 'type="page"' in element._element.xml)
"""
this is a page break, than you must keep track of your pages for this to work.

**Comment by dhdaines** [comment]
There appears to be some support for `lastRenderedPageBreak` in the `Paragraph` and `Run` classes in the form of the `contains_page_breaks` and `rendered_page_breaks` attributes

Unfortunately they only support `w:lastRenderedPageBreak` and not explicit `<w:br w:type="page"/>` which makes them less than useful at least as far as I can tell, since you can't actually rely on them to find all the page breaks in the document...

It should however be possible to count all `<w:lastRenderedPageBreak/>` and `<w:br w:type="page"/>` elements preceding a given paragraph/run/table which probably gives you a reliable notion of what the last program to edit/save the document thought the page number was.

**Comment by scanny** [comment]
Actually only counting `w:lastRenderedPageBreak` is on purpose. If you count the "hard" page breaks you'll double-count those pages because the renderer places a last-rendered break element beside those as well.

There's no good situation on this unfortunately because lastRenderedPageBreak is optional. But when they are present they seem to be pretty reliable.

You could check for rendered page-breaks and then count hard page breaks when the former aren't present, but that's not a terrifically useful indicator I wouldn't think. You can easily imagine situations where it reported 2 pages for a 100 page document.

The short answer is there is no good way to get rendered page breaks without actually rendering.

**Comment by dhdaines** [comment]
> Actually only counting `w:lastRenderedPageBreak` is on purpose. If you count the "hard" page breaks you'll double-count those pages because the renderer places a last-rendered break element beside those as well.

Indeed!  I realized this shortly after writing the comment... it's confusing since they don't end up in the same place (`w:lastRenderedPageBreak` will go immediately before the text on the new page, whereas `w:br` is wherever you put it, often in the previous paragraph)

> The short answer is there is no good way to get rendered page breaks without actually rendering.

Unfortunately yes, and once you have a PDF, you have two problems...

**Comment by dhdaines** [comment]
It would be probably helpful to have `rendered_page_breaks` defined for tables (and rows, and cells).  In theory you can obviously recurse down to the spans inside a table, but since tables can contain other tables, which can contain other tables... this is a bit of a pain when you just want to know how many pages the table spans.  In the meantime one can always just do:

```python
breaks = table._element.xpath(".//w:lastRenderedPageBreak")
```

(or likewise for rows or cells or whatever)

Once you have this, then you *should* be able to find all of the rendered page breaks.  If my reading of unstructured's code (which I think you wrote @scanny ?) is correct then it only counts page breaks inside top-level paragraphs.

I could make a PR...

**Comment by dhdaines** [comment]
Here is the definitive word on the subject of OOXML (i.e. DOCX) and page breaks: https://stackoverflow.com/questions/39992870/how-to-access-openxml-content-by-page-number/40139811#40139811

**Comment by scanny** [comment]
Closing as resolved.
